src/Dictionaries/CacheDictionary.cpp

```diff
@@ -552,16 +544,28 @@ void CacheDictionary<dictionary_key_type>::update(CacheDictionaryUpdateUnitPtr<d
         if (key_index_to_state_from_storage[i].isExpired()
             || key_index_to_state_from_storage[i].isNotFound())
         {
-            if constexpr (dictionary_key_type == DictionaryKeyType::Simple)
-                requested_keys_vector.emplace_back(requested_keys[i]);
-            else
-                requested_complex_key_rows.emplace_back(i);
 
             auto requested_key = requested_keys[i];
-            not_found_keys.insert(requested_key);
+            not_found_key_to_request_index.insert(makePairNoInit(requested_key, i));
         }
     }
 
+    std::vector<UInt64> requested_keys_vector;
+    std::vector<size_t> requested_complex_key_rows;
+
+    if constexpr (dictionary_key_type == DictionaryKeyType::Simple)
+    {
+        requested_keys_vector.reserve(requested_keys.size());
```

Comment:
not sure if it makes sense, but maybe 
``` cpp
reserve(min(requested_keys.size(), not_found_key_to_request_index.size()))
```
?

Reply:
Added simpler implementation.

Reply:
Updated implementation.